File: coding_part/wro2025_vehicle/src/control/path_planner.cpp
// control/path_planner.cpp
// Implementation of PathPlanner

#include "path_planner.h"
#include <iostream>
#include <algorithm> // For std::max, std::min

PathPlanner::PathPlanner() : lastKnownChallenge(MissionState::ChallengeType::UNKNOWN)
{
    // Constructor
}

bool PathPlanner::initialize(std::shared_ptr<LapCounter> lapCounter /*, std::shared_ptr<SignDetector> signDetector*/)
{
    if (!lapCounter /* || !signDetector */)
    {
        std::cerr << "PathPlanner: Invalid dependencies provided." << std::endl;
        return false;
    }
    this->lapCounter = lapCounter;
    // this->signDetector = signDetector;
    std::cout << "PathPlanner: Initialized with dependencies." << std::endl;
    return true;
}

NavigationCommand PathPlanner::plan(const SensorData &sensorData, const VisionSnapshot &visionData, const MissionState &missionState)
{

    // Update internal state if challenge type changed
    if (missionState.getChallengeType() != lastKnownChallenge)
    {
        lastKnownChallenge = missionState.getChallengeType();
        std::cout << "PathPlanner: Challenge type changed to "
                  << (lastKnownChallenge == MissionState::ChallengeType::OPEN_CHALLENGE ? "Open" : "Obstacle")
                  << std::endl;
    }

    // Dispatch to specific planning logic based on mission phase and challenge type
    switch (missionState.getPhase())
    {
    case MissionPhase::DRIVING_LAPS:
        if (missionState.isOpenChallenge())
        {
            return planForOpenChallenge(sensorData, visionData, missionState);
        }
        else if (missionState.isObstacleChallenge())
        {
            return planForObstacleChallenge(sensorData, visionData, missionState);
        }
        break;

    case MissionPhase::STOPPING_AT_START:
        return planForStopping(sensorData, visionData, missionState);
        break;

    case MissionPhase::FINDING_PARKING:
    case MissionPhase::PARKING:
        if (missionState.isObstacleChallenge())
        {
            return planForParking(sensorData, visionData, missionState);
        }
        break;

    case MissionPhase::EMERGENCY_STOPPED:
        return NavigationCommand(NavigationCommand::Action::EMERGENCY_STOP, 0.0f);
        break;

    // Other phases (PRE_START, WAITING_FOR_GO, COMPLETED) should not require path planning
    default:
        // Default action if not in a driving phase
        return NavigationCommand(NavigationCommand::Action::GO_STRAIGHT, 0.0f);
        break;
    }

    // Default fallback (should ideally not be reached in normal flow)
    std::cerr << "PathPlanner: No specific plan for current state. Defaulting to GO_STRAIGHT." << std::endl;
    return NavigationCommand(NavigationCommand::Action::GO_STRAIGHT, 0.5f);
}

// --- Private Helper Functions ---

NavigationCommand PathPlanner::planForOpenChallenge(const SensorData &sensorData, const VisionSnapshot &visionData, const MissionState &missionState)
{
    // In Open Challenge, there are no signs. Strategy is usually wall following.
    // Common strategies: Left Wall Following, Right Wall Following.
    // For simplicity, let's assume a basic left wall following strategy.

    // Basic obstacle avoidance check
    float turnDirection = 0.0f;
    if (needsObstacleAvoidance(sensorData, turnDirection))
    {
        float speed = 0.4f; // Slow down for obstacle
        if (turnDirection < 0)
        {
            // Turn left
            return NavigationCommand(NavigationCommand::Action::GO_STRAIGHT, speed); // Assume steering is handled by lower level
        }
        else if (turnDirection > 0)
        {
            // Turn right
            return NavigationCommand(NavigationCommand::Action::GO_STRAIGHT, speed);
        }
        else
        {
            // Stop or back up?
            return NavigationCommand(NavigationCommand::Action::GO_STRAIGHT, 0.0f);
        }
    }

    // Default left wall following logic (simplified)
    // Aim to maintain a certain distance from the left wall
    int16_t leftDistance = sensorData.leftDistance;
    int16_t frontDistance = sensorData.frontDistance;

    float desiredSpeed = 0.6f; // Medium speed
    // Action is implicit in speed/steering calculation in this simplified view
    // A more complex planner might explicitly set FOLLOW_LEFT_LANE

    // Simple proportional controller for steering based on left distance
    // Target distance from left wall (mm)
    const float TARGET_LEFT_DISTANCE = 200.0f;
    const float MAX_STEERING_ADJUSTMENT = 20.0f; // Degrees

    float distanceError = static_cast<float>(leftDistance) - TARGET_LEFT_DISTANCE;
    // Clamp error to prevent overreaction
    distanceError = std::max(-300.0f, std::min(300.0f, distanceError));

    // This logic would typically result in a servo angle command in the VehicleController
    // For example: base_angle + Kp * error
    // If error is positive (too far from wall), turn left (decrease servo angle)
    // If error is negative (too close to wall), turn right (increase servo angle)

    // For now, we just return a command indicating intent and speed
    // The actual steering calculation happens in VehicleController/Actuators
    return NavigationCommand(NavigationCommand::Action::GO_STRAIGHT, desiredSpeed);
}

NavigationCommand PathPlanner::planForObstacleChallenge(const SensorData &sensorData, const VisionSnapshot &visionData, const MissionState &missionState)
{
    // In Obstacle Challenge, follow traffic signs.
    // Green sign means follow left lane.
    // Red sign means follow right lane.

    // Basic obstacle avoidance check (takes precedence)
    float turnDirection = 0.0f;
    if (needsObstacleAvoidance(sensorData, turnDirection))
    {
        float speed = 0.4f;
        return NavigationCommand(NavigationCommand::Action::GO_STRAIGHT, speed);
    }

    // Determine lane to follow based on sign detection
    // This assumes VisionSnapshot has been processed to detect signs
    bool followLeft = false;
    bool followRight = false;

    // Simplified logic: if both signs detected, prioritize based on last known rule or proximity
    // If only one sign, follow that rule.
    // If no signs, default to a safe strategy (e.g., center, or last known rule)

    if (visionData.signLeftDetected && visionData.signRightDetected)
    {
        // Conflict resolution needed. Simple approach: follow the closer one.
        // Assuming centroid X coordinates are available in VisionSnapshot
        // (This requires updating VisionSnapshot struct and image_processor)
        // For now, let's assume a simple flag or last known state is used.
        // Placeholder logic:
        followLeft = true; // Default to left if conflict, or use more sophisticated logic
        // std::cout << "PathPlanner: Conflict - Both signs detected. Following default (Left)." << std::endl;
    }
    else if (visionData.signLeftDetected)
    {
        followLeft = true;
        // std::cout << "PathPlanner: Green sign detected. Following left lane." << std::endl;
    }
    else if (visionData.signRightDetected)
    {
        followRight = true;
        // std::cout << "PathPlanner: Red sign detected. Following right lane." << std::endl;
    }
    else
    {
        // No signs detected. Continue with last known strategy or default.
        // This is a common scenario between signs.
        // std::cout << "PathPlanner: No signs detected. Continuing last strategy or default." << std::endl;
        // For now, we'll assume a default action like GO_STRAIGHT
        // A more robust system might remember the last sign or use path prediction.
    }

    float desiredSpeed = 0.6f; // Medium speed

    if (followLeft)
    {
        return NavigationCommand(NavigationCommand::Action::FOLLOW_LEFT_LANE, desiredSpeed);
    }
    else if (followRight)
    {
        return NavigationCommand(NavigationCommand::Action::FOLLOW_RIGHT_LANE, desiredSpeed);
    }
    else
    {
        // No specific lane command, go straight
        return NavigationCommand(NavigationCommand::Action::GO_STRAIGHT, desiredSpeed);
    }
}

NavigationCommand PathPlanner::planForStopping(const SensorData &sensorData, const VisionSnapshot &visionData, const MissionState &missionState)
{
    // This phase is for stopping in the start section after 3 laps.
    // The main logic for *when* to stop should be in main.cpp using PositionalMemory.
    // The PathPlanner's role here might be to refine the approach or execute the final stop.

    // Indicate to the control system that stopping is the goal.
    // The actual stopping maneuver (speed=0, precise positioning) is handled by VehicleController
    // based on this command and potentially further input from PositionalMemory/Vision.

    // For now, simply command a stop.
    // A more advanced version might include a "slow_approach" action.
    std::cout << "PathPlanner: Planning for STOP in start section." << std::endl;
    return NavigationCommand(NavigationCommand::Action::STOP_IN_START, 0.0f);
}

NavigationCommand PathPlanner::planForParking(const SensorData &sensorData, const VisionSnapshot &visionData, const MissionState &missionState)
{
    // Handle parking phases
    if (missionState.getPhase() == MissionPhase::FINDING_PARKING)
    {
        // Logic to search for the parking lot entrance.
        // This could involve:
        // - Looking for a specific gap in the outer wall (using vision/US)
        // - Following the outer wall until the gap is found
        // - Entering a state where the vehicle is aligned with the gap

        // Simplified placeholder: Assume we are looking for a large opening on the right
        // indicated by a sudden increase in right sensor distance.
        // A real implementation would be more robust.
        if (sensorData.rightDistance > 400)
        { // Example threshold for gap
            std::cout << "PathPlanner: Parking entrance potentially found (large right distance)." << std::endl;
            // This would likely trigger a state change in MissionState to PARKING
            // and a specific parking maneuver command.
            // For now, we can signal readiness.
            return NavigationCommand(NavigationCommand::Action::EXECUTE_PARKING, 0.3f); // Slow approach speed
        }
        else
        {
            // Continue searching, e.g., follow outer wall
            return NavigationCommand(NavigationCommand::Action::FOLLOW_RIGHT_LANE, 0.5f);
        }
    }
    else if (missionState.getPhase() == MissionPhase::PARKING)
    {
        // Execute the specific parallel parking maneuver.
        // This is a complex sequence of movements.
        // The command here signals that the complex maneuver should be executed.
        // The actual sequence of motor/servo commands would be handled by a dedicated
        // ParkingAssistant module or state machine within VehicleController.
        std::cout << "PathPlanner: Executing PARKING maneuver." << std::endl;
        return NavigationCommand(NavigationCommand::Action::EXECUTE_PARKING, 0.0f); // Speed controlled internally by parking sequence
    }

    // Default fallback for parking phases
    return NavigationCommand(NavigationCommand::Action::GO_STRAIGHT, 0.0f);
}

bool PathPlanner::needsObstacleAvoidance(const SensorData &sensorData, float &turnDirection)
{
    turnDirection = 0.0f; // Default: no turn needed

    // Check front sensor for immediate obstacles
    if (sensorData.frontDistance < 150)
    { // e.g., 15cm
        // Obstacle ahead, need to decide turn direction
        // Use left/right sensors to decide
        if (sensorData.leftDistance > sensorData.rightDistance)
        {
            turnDirection = -1.0f; // Turn left
        }
        else
        {
            turnDirection = 1.0f; // Turn right
        }
        return true;
    }

    // Optional: Check side sensors for very close walls that might require correction
    // This is more for fine-tuning path than major avoidance.
    // Example:
    // if (sensorData.leftDistance < 50) { turnDirection = 1.0f; return true; }
    // if (sensorData.rightDistance < 50) { turnDirection = -1.0f; return true; }

    return false; // No immediate obstacle avoidance needed
}

File: coding_part/wro2025_vehicle_python/control/path_planner.py
# control/path_planner.py
"""Advanced Path Planner, mirroring the C++ structure and logic."""

import logging
from typing import TYPE_CHECKING
from navigation.mission_state import MissionState, MissionPhase
# Import our data structures
from control.navigation_command import NavigationCommand, NavigationAction
# Use TYPE_CHECKING to avoid circular imports during type checking
if TYPE_CHECKING:
    from sensors.sensor_data import SensorData
    from vision.vision_snapshot import VisionSnapshot
    from navigation.mission_state import MissionState
    from navigation.lap_counter import LapCounter
    # from vision.sign_detector import SignDetector # If we had a separate sign detector

logger = logging.getLogger(__name__)

class PathPlanner:
    """
    Plans high-level navigation commands based on sensor data, vision data, and mission state.
    Mirrors the C++ PathPlanner class structure.
    """

    def __init__(self):
        """Initialize the PathPlanner."""
        self.last_known_challenge = None # To track changes
        # Dependencies (to be injected or initialized later)
        self.lap_counter: 'LapCounter' = None # Placeholder
        # self.sign_detector = None # If separate

        logger.info("PathPlanner initialized.")

    def initialize(self, lap_counter: 'LapCounter'): # , sign_detector=None): # Add sign_detector if needed
        """
        Initialize the PathPlanner with its dependencies.
        Args:
            lap_counter: An instance of LapCounter.
            # sign_detector: An instance of SignDetector (optional).
        Returns:
            bool: True if initialization is successful.
        """
        if not lap_counter: # or not sign_detector):
            logger.error("PathPlanner: Invalid dependencies provided.")
            return False

        self.lap_counter = lap_counter
        # self.sign_detector = sign_detector
        logger.info("PathPlanner: Initialized with dependencies.")
        return True

    def plan(self, sensor_data: 'SensorData', vision_snapshot: 'VisionSnapshot', mission_state: 'MissionState') -> NavigationCommand:
        """
        Main planning function. Dispatches to specific planning methods based on mission state.
        Args:
            sensor_data: Latest sensor data.
            vision_snapshot: Latest vision processing results.
            mission_state: Current mission state.
        Returns:
            NavigationCommand: The planned command.
        """
        # Update internal state if challenge type changed
        current_challenge = mission_state.get_challenge_type()
        logger.debug(f"PathPlanner.plan() called with challenge type: {current_challenge}") # Add this line for debugging
        if current_challenge != self.last_known_challenge:
            self.last_known_challenge = current_challenge
            challenge_name = "Open" if mission_state.is_open_challenge() else "Obstacle"
            logger.info(f"PathPlanner: Challenge type changed to {challenge_name}")

        # Dispatch to specific planning logic based on mission phase and challenge type
        phase = mission_state.get_phase()
        try:
            if phase == MissionPhase.DRIVING_LAPS:
                if mission_state.is_open_challenge():
                    return self._plan_for_open_challenge(sensor_data, vision_snapshot, mission_state)
                elif mission_state.is_obstacle_challenge():
                    return self._plan_for_obstacle_challenge(sensor_data, vision_snapshot, mission_state)

            elif phase in (MissionPhase.STOPPING_AT_START,):
                # This phase is for stopping in the start section after 3 laps.
                # The main logic for *when* to stop should be in main loop using PositionalMemory.
                # PathPlanner provides the command to execute the stop.
                return self._plan_for_stopping(sensor_data, vision_snapshot, mission_state)

            elif phase in (MissionPhase.FINDING_PARKING, MissionPhase.PARKING):
                if mission_state.is_obstacle_challenge():
                    return self._plan_for_parking(sensor_data, vision_snapshot, mission_state)

            elif phase == MissionPhase.EMERGENCY_STOPPED:
                return NavigationCommand(NavigationAction.EMERGENCY_STOP, 0.0)

            else:
                # Phases like PRE_START, WAITING_FOR_GO, COMPLETED don't require path planning here
                # Default action if no specific logic applies
                logger.debug(f"PathPlanner: No specific plan for phase {phase}. Defaulting to GO_STRAIGHT.")
                return NavigationCommand(NavigationAction.GO_STRAIGHT, 0.0)

        except Exception as e:
            logger.error(f"PathPlanner: Error during planning: {e}", exc_info=True)
            # Fail-safe: Emergency stop on planning error
            return NavigationCommand(NavigationAction.EMERGENCY_STOP, 0.0)

    def _plan_for_open_challenge(self, sensor_data: 'SensorData', vision_snapshot: 'VisionSnapshot', mission_state: 'MissionState') -> NavigationCommand:
        """
        Plan path for the Open Challenge (no traffic signs, follow walls/center line).
        Args:
            sensor_data: Latest sensor data.
            vision_snapshot: Latest vision processing results.
            mission_state: Current mission state.
        Returns:
            NavigationCommand: The planned command.
        """
        logger.debug("PathPlanner: Planning for Open Challenge.")
        desired_speed = 0.6 # Base speed

        # --- Basic Obstacle Avoidance (Fallback/Supplement) ---
        # This takes precedence over normal navigation
        avoidance_needed, turn_direction = self._needs_obstacle_avoidance(sensor_data)
        if avoidance_needed:
            avoidance_speed = 0.4 # Slow down
            if turn_direction < 0: # Turn left
                logger.info("PathPlanner (Open): Obstacle avoidance - turning left.")
                # For simplicity, assume steering is handled by lower level based on US
                return NavigationCommand(NavigationAction.GO_STRAIGHT, avoidance_speed)
            elif turn_direction > 0: # Turn right
                logger.info("PathPlanner (Open): Obstacle avoidance - turning right.")
                return NavigationCommand(NavigationAction.GO_STRAIGHT, avoidance_speed)
            else: # Go straight (e.g., center obstacle, back up?)
                 logger.info("PathPlanner (Open): Obstacle avoidance - going straight/backing up?.")
                 return NavigationCommand(NavigationAction.GO_STRAIGHT, -avoidance_speed) # Negative for reverse?

        # --- Wall Following / Center Line Following Logic ---
        # Use vision data (edge detection) or sensor data (US distances) for wall following.
        # Simple Proportional control based on vision edge data
        left_edge_x = vision_snapshot.avg_left_edge_x
        right_edge_x = vision_snapshot.avg_right_edge_x
        track_width_estimate = vision_snapshot.get_track_width_estimate()

        # Assume camera resolution width is known (from config)
        import config.vehicle_config as cfg
        frame_center_x = cfg.CAMERA_WIDTH / 2.0

        # Strategy: Follow center of track
        if left_edge_x >= 0 and right_edge_x >= 0 and track_width_estimate > 50: # Valid edges
            track_center_x = (left_edge_x + right_edge_x) / 2.0
            error = track_center_x - frame_center_x

            # Simple P controller (Kp needs tuning)
            Kp = 0.01 # Tuning parameter
            steering_adjustment = -Kp * error # Negative because positive error means turn right

            # Map steering_adjustment to servo angle or command
            # This is conceptual; actual mapping happens in VehicleController
            # For command, we might just indicate intent and speed
            # A more advanced system might calculate a specific steering angle/servo command here

            if abs(error) < 20: # Within tolerance
                logger.debug(f"PathPlanner (Open): Centered (Error: {error:.1f}). Going straight.")
                return NavigationCommand(NavigationAction.GO_STRAIGHT, desired_speed)
            elif error > 0: # Too far right, need to turn left
                logger.debug(f"PathPlanner (Open): Off-center (Error: {error:.1f}). Turning left.")
                return NavigationCommand(NavigationAction.TURN_LEFT, desired_speed)
            else: # error < 0, Too far left, need to turn right
                logger.debug(f"PathPlanner (Open): Off-center (Error: {error:.1f}). Turning right.")
                return NavigationCommand(NavigationAction.TURN_RIGHT, desired_speed)

        elif left_edge_x >= 0: # Only left edge visible, follow left wall
             logger.debug("PathPlanner (Open): Following left wall.")
             return NavigationCommand(NavigationAction.FOLLOW_LEFT_LANE, desired_speed)
        elif right_edge_x >= 0: # Only right edge visible, follow right wall
             logger.debug("PathPlanner (Open): Following right wall.")
             return NavigationCommand(NavigationAction.FOLLOW_RIGHT_LANE, desired_speed)
        else:
            # No clear edges, go straight cautiously
            logger.warning("PathPlanner (Open): No clear track edges detected. Proceeding straight.")
            return NavigationCommand(NavigationAction.GO_STRAIGHT, desired_speed * 0.8) # Slow down

    def _plan_for_obstacle_challenge(self, sensor_data: 'SensorData', vision_snapshot: 'VisionSnapshot', mission_state: 'MissionState') -> NavigationCommand:
        """
        Plan path for the Obstacle Challenge (follow traffic signs).
        Args:
            sensor_data: Latest sensor data.
            vision_snapshot: Latest vision processing results.
            mission_state: Current mission state.
        Returns:
            NavigationCommand: The planned command.
        """
        logger.debug("PathPlanner: Planning for Obstacle Challenge.")
        desired_speed = 0.6 # Base speed

        # --- Basic Obstacle Avoidance (Fallback/Supplement) ---
        # Takes precedence even in Obstacle Challenge
        avoidance_needed, turn_direction = self._needs_obstacle_avoidance(sensor_data)
        if avoidance_needed:
             avoidance_speed = 0.4
             if turn_direction < 0:
                 logger.info("PathPlanner (Obstacle): Obstacle avoidance - turning left.")
                 return NavigationCommand(NavigationAction.GO_STRAIGHT, avoidance_speed)
             elif turn_direction > 0:
                 logger.info("PathPlanner (Obstacle): Obstacle avoidance - turning right.")
                 return NavigationCommand(NavigationAction.GO_STRAIGHT, avoidance_speed)
             else:
                 logger.info("PathPlanner (Obstacle): Obstacle avoidance - going straight.")
                 return NavigationCommand(NavigationAction.GO_STRAIGHT, -avoidance_speed)

        # --- Sign Following Logic ---
        # Determine lane to follow based on sign detection
        primary_left_sign = vision_snapshot.primary_left_sign
        primary_right_sign = vision_snapshot.primary_right_sign

        follow_left = False
        follow_right = False

        # Logic: if both signs detected with high confidence, it's a conflict or close proximity.
        # Prioritize based on confidence or last known rule (e.g., last sign seen?).
        # For simplicity, let's prioritize the one with higher confidence, or left if equal/both low.
        # A more robust system would need better conflict resolution.

        if primary_left_sign and primary_left_sign.is_valid() and primary_left_sign.confidence > 0.5:
            follow_left = True
        if primary_right_sign and primary_right_sign.is_valid() and primary_right_sign.confidence > 0.5:
            follow_right = True

        if follow_left and follow_right:
            # Conflict: Both signs detected. Resolve.
            logger.warning("PathPlanner (Obstacle): Potential sign conflict detected.")
            if primary_left_sign.confidence > primary_right_sign.confidence:
                follow_right = False
                logger.info("PathPlanner (Obstacle): Resolved conflict, following LEFT sign (higher confidence).")
            elif primary_right_sign.confidence > primary_left_sign.confidence:
                follow_left = False
                logger.info("PathPlanner (Obstacle): Resolved conflict, following RIGHT sign (higher confidence).")
            else:
                # Equal confidence or both low, default to left or last known direction?
                # Let's default to left for now.
                follow_right = False
                logger.info("PathPlanner (Obstacle): Resolved conflict, defaulting to LEFT sign.")

        if follow_left:
            logger.info("PathPlanner (Obstacle): Following LEFT lane as instructed by sign.")
            return NavigationCommand(NavigationAction.FOLLOW_LEFT_LANE, desired_speed)
        elif follow_right:
            logger.info("PathPlanner (Obstacle): Following RIGHT lane as instructed by sign.")
            return NavigationCommand(NavigationAction.FOLLOW_RIGHT_LANE, desired_speed)
        else:
            # No clear sign instruction. Default behavior?
            # Could be: follow center, follow last known direction, or wait/slow down.
            # Let's assume follow center line based on edges if available, otherwise slow down.
            logger.info("PathPlanner (Obstacle): No clear sign instruction. Following center line or slowing down.")
            # Re-use open challenge centering logic as fallback
            return self._plan_for_open_challenge(sensor_data, vision_snapshot, mission_state) # Or a simplified version


    def _plan_for_stopping(self, sensor_data: 'SensorData', vision_snapshot: 'VisionSnapshot', mission_state: 'MissionState') -> NavigationCommand:
        """
        Plan for stopping in the start section (Open) or finishing (Obstacle).
        The decision to stop is made externally (e.g., by PositionalMemory in main loop).
        This just provides the command to execute the stop.
        Args:
            sensor_data: Latest sensor data.
            vision_snapshot: Latest vision processing results.
            mission_state: Current mission state.
        Returns:
            NavigationCommand: The STOP command.
        """
        logger.info("PathPlanner: Generating STOP command for stopping phase.")
        # The action (STOP_IN_START) implies stopping and potentially specific maneuver
        # The speed is 0.0
        return NavigationCommand(NavigationAction.STOP_IN_START, 0.0)

    def _plan_for_parking(self, sensor_data: 'SensorData', vision_snapshot: 'VisionSnapshot', mission_state: 'MissionState') -> NavigationCommand:
        """
        Plan for finding and executing parking (Obstacle Challenge).
        """
        # ... (previous code) ...

        # --- Refine Parking Logic ---
        if phase == mission_state.MissionPhase.FINDING_PARKING:
            logger.debug("PathPlanner: In FINDING_PARKING phase.")
            # Simplified check: Assume we are in the start section and looking for the spot.
            # A more robust check would use positional memory or vision to confirm proximity to the parking area.
            # The C++ code has a placeholder check for large right distance.
            # Let's use a similar simple check for now, but acknowledge positional memory exists.
            # In a full implementation, PathPlanner might receive a reference to PositionalMemory.

            # Example check (replace with more robust logic using PositionalMemory if available in scope)
            # if mission_state.is_in_start_section(): # This flag needs setting logic
            if sensor_data.us_data.right_distance > 100: # Placeholder for "gap detected"
                logger.info("PathPlanner (Parking): Parking entrance potentially found. Initiating parking sequence.")
                return NavigationCommand(NavigationAction.EXECUTE_PARKING, 0.0)
            else:
                logger.debug("PathPlanner (Parking): Navigating towards parking area.")
                # Could follow outer wall or re-trace path.
                return NavigationCommand(NavigationAction.FOLLOW_RIGHT_LANE, 0.4) # Slow approach

        elif phase == mission_state.MissionPhase.PARKING:
            logger.debug("PathPlanner: In PARKING phase.")
            # Execute the specific parking maneuver command.
            # The actual sequence is handled by VehicleController.
            logger.info("PathPlanner: Executing PARKING maneuver command.")
            return NavigationCommand(NavigationAction.EXECUTE_PARKING, 0.0)


    def _needs_obstacle_avoidance(self, sensor_data: 'SensorData') -> tuple[bool, float]:
        """
        Basic obstacle avoidance logic based on ultrasonic sensors.
        Args:
            sensor_data: Latest sensor data.
        Returns:
            tuple[bool, float]: (needs_avoidance, turn_direction).
                            turn_direction: -1=left, 1=right, 0=none/straight.
        """
        # Thresholds (these are examples, need tuning)
        front_threshold = 20.0  # cm
        side_threshold = 15.0   # cm

        front_dist = sensor_data.us_data.front_distance
        left_dist = sensor_data.us_data.left_distance
        right_dist = sensor_data.us_data.right_distance
        # back_dist = sensor_data.us_data.back_distance # Might be useful for backing up

        if front_dist < front_threshold and front_dist > 0: # Valid reading and too close
            # Obstacle in front
            if left_dist > right_dist and left_dist > side_threshold:
                # More space on the left
                logger.debug(f"ObstacleAvoidance: Front obstacle. Turning LEFT. (F:{front_dist:.1f}, L:{left_dist:.1f}, R:{right_dist:.1f})")
                return True, -1.0 # Turn left
            elif right_dist > left_dist and right_dist > side_threshold:
                # More space on the right
                logger.debug(f"ObstacleAvoidance: Front obstacle. Turning RIGHT. (F:{front_dist:.1f}, L:{left_dist:.1f}, R:{right_dist:.1f})")
                return True, 1.0 # Turn right
            else:
                # No clear side preference or boxed in
                logger.debug(f"ObstacleAvoidance: Front obstacle. No clear side or boxed in. (F:{front_dist:.1f}, L:{left_dist:.1f}, R:{right_dist:.1f})")
                return True, 0.0 # Go straight/back up?

        # Could also check for side obstacles forcing a turn away
        # e.g., if left_dist < side_threshold and right_dist > left_dist + 5:
        #       return True, 1.0 # Turn right to avoid left wall

        # No immediate front obstacle requiring action
        return False, 0.0


# Example usage (if run as script)
if __name__ == "__main__":
    import utils.logger
    utils.logger.setup_logging()

    # This would normally be integrated into the main application flow
    # and receive real data objects.
    planner = PathPlanner()
    # Mock dependencies would be needed to run this standalone
    print("PathPlanner class defined.")
